### PR TITLE
opm serve: use pre-existing cache, if set and up-to-date

### PR DESCRIPTION
**Pull request by joelanford**
Signed-off-by: Joe Lanford <joe.lanford@gmail.com>

**Description of the change:**
This PR introduces an optional persistent caching layer to the `opm serve` command to reduce the cost of initializing the GRPC server for file-based catalogs -- and thus improve the startup time.

It also update the Dockerfile generated by `opm generate dockerfile` to build a cache at image build time and use it at runtime. This does have the side affect of essentially doubling the image size when compared to an image that does not pre-generate the cache.

It adds `--cache-dir` and `--cache-only` flags to the `opm serve` command to enable interactions with the cache.
1. The `--cache-dir` flag specifies a persistent cache directory to use. If `--cache-dir` is unset, the existing behavior (which builds and uses an ephemeral cache directory) remains.
2. The `--cache-only` flag causes `opm serve` to sync the cache and then exit without actually serving the GRPC API. This is useful when building catalog images -- if the catalog image contains a pre-synced cache, the container startup time is significantly improved.

**Motivation for the change:**
The `opm serve` command's startup time is proportional to the size of the FBC. This is a problem because OLM creates catalog pods with somewhat aggressive health/liveness probes that cause affected FBC catalogs to never become healthy.

This problem has been partially addressed in current/future OLM versions via the introduction of a startupProbe. However, that is mainly a workaround for the root problem. This PR attempts to improve the startup time itself.

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/registry/query.go [hunk 1]
```diff
@@ -392,24 +355,245 @@ func channelEntriesThatReplace(b model.Bundle, name string) []*ChannelEntry {
...
+	Name    string
+	Head    string
+	Bundles map[string]cBundle
+}
+
+type cBundle struct {
+	Package  string   `json:"package"`
+	Channel  string   `json:"channel"`
+	Name     string   `json:"name"`
+	Replaces string   `json:"replaces""`
```
Comment (grokspawn) [inline pkg/registry/query.go:597]:
```suggestion
	Replaces string   `json:"replaces"`
```

File: pkg/registry/tar.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"time"
+)
+
+// fsToTarGZ writes the filesystem represented by fsys to w as a gzipped tar archive.
```
Comment (grokspawn) [inline pkg/registry/tar.go:12]: Is this still accurate?  Method name mismatch, and I didn't see any compression.

File: pkg/registry/query.go [hunk 3]
```diff
@@ -392,24 +385,254 @@ func channelEntriesThatReplace(b model.Bundle, name string) []*ChannelEntry {
...
+		return nil, err
+	}
+	qc := &cache{baseDir: baseDir, persist: true}
+	if digest, err := os.ReadFile(filepath.Join(baseDir, "digest")); err == nil {
+		qc.digest = strings.TrimSpace(string(digest))
+	}
+	return qc, nil
+}
+
+func (qc *cache) load(getDigest func() (string, error), getModel func() (model.Model, error)) error {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply pkg/registry/query.go:467]: This is what I meant by my slack comment of "this is pretty ugly"

File: pkg/registry/query.go [hunk 4]
```diff
@@ -392,24 +385,254 @@ func channelEntriesThatReplace(b model.Bundle, name string) []*ChannelEntry {
...
+
+type cache struct {
+	digest     string
+	baseDir    string
+	persist    bool
+	pkgs       map[string]cPkg
+	apiBundles map[apiBundleKey]string
+}
+
+func newCache(baseDir string, getDigest func() (string, error), getModel func() (model.Model, error)) (*cache, error) {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply pkg/registry/query.go:418]: The idea behind the model-only based querier was originally: "we don't care what the source format is, we can always serve the grpc API if you can convert it to the model first"

I'm not totally ready to give up on that idea, but once sqlite support disappears, the only format that even exists is FBC.

[the remark this answers is not in the document]
Comment (grokspawn) [reply pkg/registry/query.go:418]: I'm definitely not arguing about it right now.  

My only concern is that it may be difficult to recognize this as a dead path with the sunset of SQLite indices.  I figured someone will do a 'grep -i sqlite' over the code body, and it would stand out if there were a comment.

File: pkg/registry/query.go [hunk 5]
```diff
@@ -392,24 +385,254 @@ func channelEntriesThatReplace(b model.Bundle, name string) []*ChannelEntry {
...
+				newCh.Bundles[b.Name] = newB
+			}
+			newP.Channels[ch.Name] = newCh
+		}
+		pkgs[p.Name] = newP
+	}
+	return pkgs, nil
+}
+
+type cPkg struct {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply pkg/registry/query.go:618]: These are actually more analogous to the model.Model types (just slimmed down)

For the most part, I think `packagesFromModel` will give us compile-time "hey, you removed something from the model" checking.

And if we _add_ some fields to the model, I think that'll mostly be changes that need to be made in `api.ConvertModelBundleToAPIBundle(*b)`

There's probably an edge case or two that we could miss, but I'd hope that if we change something that would require touching these types, we'd have a test fail somewhere.

File: pkg/registry/query.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joelanford) [reply pkg/registry/query.go:418]: `NewQuerier` isn't actually used in any of the sqlite flows as far as I know. At this point, it is only used in tests, where its much easier and compile-time safe to create a querier from a `model.Model` than it is to build an fs.FS containing FBC.

With that in mind, I'm guessing we'll keep NewQuerier around even after sqlite support is completely removed.

**Comment by jdockter** [comment]
@joelanford was able to test your branch by building `opm` binary local and pulling based image `quay.io/joelanford/opm:opm-serve-cache-personal` in FBC build of the IBM Operator catalog.  Tested internally on OCP 4.11 x, p, z cluster as well as OCP 4.6 x and p with no issue.  Saw startup times of catalog pod around 15 seconds using updated dockerfile for catalog build like the following:

```
FROM --platform=linux/amd64 quay.io/joelanford/opm:opm-serve-cache-personal AS builder
FROM --platform=linux/amd64 registry.redhat.io/ubi8/ubi-minimal
COPY --from=builder /bin/opm/ /bin/opm
COPY --from=builder /bin/grpc_health_probe /bin/grpc_health_probe
### add licenses to this directory
RUN mkdir /licenses
COPY licenses/LICENSE /licenses
WORKDIR /tmp
ARG user=1001
RUN microdnf clean all && chmod -vR g=u /etc/passwd
USER ${user}
ENTRYPOINT ["/bin/opm"]
CMD ["serve", "/configs", "--cache-dir=/tmp/cache"]
COPY amd64 /configs
RUN ["/bin/opm", "serve", "/configs", "--cache-dir=/tmp/cache", "--cache-only"]
LABEL operators.operatorframework.io.index.configs.v1=/configs
```
